src/components/action-bar/action-bar.tsx

```diff
@@ -0,0 +1,275 @@
+    shadow: true,
+    styleUrl: 'action-bar.scss',
+})
+export class ActionBar {
+    /**
+     * Items that are placed in the action bar.
+     * These represent primary actions.
+     */
+    @Prop()
+    public actionBarItems: Array<ActionBarItem | ListSeparator> = [];
+
+    /**
+     * A label used to describe the purpose of the element to users
+     * of assistive technologies, like screen readers.
+     * Example value: "toolbar"
+     */
+    @Prop({ reflect: true })
+    public accessibleLabel?: string;
+
+    /**
+     * - When set to `fullWidth`, the component will take the
+     * entire width of its container.
+     * - When set to `floating`, the component will get basic stylings
+     * to visualize the floating state.
+     * :::note
+     * You should still properly position the component
+     * according to the structure of your user interface.
+     * For example, use an `absolute` or `fixed` position.
+     * :::
+     */
+    @Prop({ reflect: true })
+    public layout?: 'fullWidth' | 'floating';
+
+    /**
+     * Fired when a action bar item has been clicked.
+     */
+    @Event()
+    public itemSelected: EventEmitter<ActionBarItem>;
+
+    @Element()
+    public element: HTMLElement;
+
+    /**
+     * Defines the location that the content of the overflow menu
+     * appears, in relation to its trigger.
+     */
+    @Prop({ reflect: true })
+    public openDirection: OpenDirection;
+
+    @State()
+    private overflowCutoff: number = this.actionBarItems.length;
+
+    @State()
+    private calculatingOverflow: boolean = false;
+
+    private resizeObserver: ResizeObserver;
+    private intersectionObserver: IntersectionObserver;
+    private overflowRecalculationScheduled: boolean;
+    // eslint-disable-next-line no-magic-numbers
+    private RESIZE_RATE_MS = 50;
+    // eslint-disable-next-line no-magic-numbers
+    private RESIZE_TIMEOUT = 250;
+    private lastOverflowCutoff: number = this.actionBarItems.length;
+    private lastResize = Date.now();
+
+    public componentWillLoad() {
+        this.resizeObserver = new ResizeObserver(this.handleResize);
+    }
+
+    disconnectedCallback() {
+        this.resizeObserver.disconnect();
+        this.intersectionObserver.disconnect();
+    }
+
+    public render() {
+        const visibleItems = this.actionBarItems.slice(0, this.overflowCutoff);
+        const overflowItems = this.actionBarItems.slice(this.overflowCutoff);
+
+        return (
+            <Host
+                aria-label={this.accessibleLabel}
+                class={{
+                    'is-full-width': this.layout === 'fullWidth',
+                    'is-floating': this.layout === 'floating',
+                }}
+            >
+                {visibleItems.map(this.renderActionBarItem)}
+                {this.renderOverflowMenu(overflowItems)}
+            </Host>
+        );
+    }
+
+    public componentDidRender() {
+        const element = this.element.parentElement;
```

Comment:
This will not work if the component is an immediate child of a web-component with a shadow DOM. In those cases the parent element can be fetched by
```ts
element.parentNode.host
```

Reply:
Oh this is really nifty! You are right in your analysis when the problem occur. Let's try this out @LawrenceBorst

Reply:
if we do so, we need to update the docs too. just keep in mind